proptools/filter.go

```diff
@@ -0,0 +1,158 @@
+	for _, field := range fields {
+		var keep bool
+		if keep, field = predicate(field, prefix); !keep {
+			filtered = true
+			continue
+		}
+
+		subPrefix := field.Name
+		if prefix != "" {
+			subPrefix = prefix + "." + subPrefix
+		}
+
+		// Recurse into structs
+		switch field.Type.Kind() {
+		case reflect.Struct:
+			var subFiltered bool
+			field.Type, subFiltered = filterPropertyStruct(field.Type, subPrefix, predicate)
+			filtered = filtered || subFiltered
+			if field.Type == nil {
+				continue
+			}
+		case reflect.Ptr:
+			if field.Type.Elem().Kind() == reflect.Struct {
+				nestedType, subFiltered := filterPropertyStruct(field.Type.Elem(), subPrefix, predicate)
+				filtered = filtered || subFiltered
+				if nestedType == nil {
+					continue
+				}
+				field.Type = reflect.PtrTo(nestedType)
+			}
+		case reflect.Interface:
+			panic("Interfaces are not supported in arch_variant properties")
+		}
+
+		filteredFields = append(filteredFields, field)
+	}
+
+	return filteredFields, filtered
+}
+
+// FilterPropertyStruct takes a reflect.Type that is either a struct or a pointer to a struct, and returns a
+// reflect.Type that only contains the fields in the original type for which predicate returns true, and a bool
+// that is true if the new struct type has fewer fields than the original type.  If there are no fields in the
+// original type for which predicate returns true it returns nil and true.
+func FilterPropertyStruct(prop reflect.Type, predicate FilterFieldPredicate) (filteredProp reflect.Type, filtered bool) {
+	return filterPropertyStruct(prop, "", predicate)
+}
+
+func filterPropertyStruct(prop reflect.Type, prefix string, predicate FilterFieldPredicate) (filteredProp reflect.Type, filtered bool) {
+	var fields []reflect.StructField
+
+	ptr := prop.Kind() == reflect.Ptr
+	if ptr {
+		prop = prop.Elem()
+	}
+
+	for i := 0; i < prop.NumField(); i++ {
+		fields = append(fields, prop.Field(i))
+	}
+
+	filteredFields, filtered := filterPropertyStructFields(fields, prefix, predicate)
+
+	if len(filteredFields) == 0 {
+		return nil, true
+	}
+
+	if !filtered {
+		if ptr {
+			return reflect.PtrTo(prop), false
+		}
+		return prop, false
+	}
+
+	ret := reflect.StructOf(filteredFields)
+	if ptr {
+		ret = reflect.PtrTo(ret)
+	}
+
+	return ret, true
+}
+
+// FilterPropertyStructSharded takes a reflect.Type that is either a sturct or a pointer to a struct, and returns a list
+// of reflect.Type that only contains the fields in the original type for which predicate returns true, and a bool that
+// is true if the new struct type has fewer fields than the original type.  If there are no fields in the original type
+// for which predicate returns true it returns nil and true.  Each returned struct type will have a maximum of 10 top
+// level fields in it to attempt to avoid hitting the reflect.StructOf name length limit, although the limit can still
```

Comment:
Q: What are you referring to by "reflect.StructOf name length limit"? My quick Google search failed to find any relevant info.

Reply:
Reworded, it hits a panic at https://golang.org/src/reflect/type.go#L537.